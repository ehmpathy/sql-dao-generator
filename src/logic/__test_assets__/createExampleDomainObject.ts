import { DomainObjectMetadata, DomainObjectVariant } from 'domain-objects-metadata';

export const createExampleDomainObjectMetadata = ({
  extend = DomainObjectVariant.DOMAIN_VALUE_OBJECT,
}: {
  extend?: DomainObjectVariant;
} = {}) => {
  if (extend === DomainObjectVariant.DOMAIN_VALUE_OBJECT)
    return new DomainObjectMetadata({
      name: 'ExampleDomainValueObject',
      extends: DomainObjectVariant.DOMAIN_VALUE_OBJECT,
      properties: {},
      decorations: { unique: null, updatable: null },
    });

  if (extend === DomainObjectVariant.DOMAIN_ENTITY)
    return new DomainObjectMetadata({
      name: 'ExampleDomainEntity',
      extends: DomainObjectVariant.DOMAIN_VALUE_OBJECT,
      properties: {},
      decorations: { unique: ['uuid'], updatable: [] },
    });
  if (extend === DomainObjectVariant.DOMAIN_EVENT)
    return new DomainObjectMetadata({
      name: 'ExampleDomainEvent',
      extends: extend,
      properties: {},
      decorations: { unique: ['uuid'], updatable: null },
    });
  throw new Error('unsupported domain object variant to create example object for');
};
