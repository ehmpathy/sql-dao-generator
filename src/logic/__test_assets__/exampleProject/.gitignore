provision/schema/declarations/**/*
provision/schema/sql/**/*
provision/schema/sql/**/*
src/data/dao
