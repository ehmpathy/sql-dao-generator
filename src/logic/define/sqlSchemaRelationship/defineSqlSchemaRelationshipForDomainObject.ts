import { snakeCase } from 'change-case';
import {
  DomainObjectMetadata,
  DomainObjectPropertyType,
  DomainObjectVariant,
} from 'domain-objects-metadata';

import { SqlSchemaToDomainObjectRelationship } from '../../../domain/objects/SqlSchemaToDomainObjectRelationship';
import { UnexpectedCodePathDetectedError } from '../../UnexpectedCodePathDetectedError';
import { UserInputError } from '../../UserInputError';
import { assertDomainObjectIsSafeToHaveSqlSchema } from './assertDomainObjectIsSafeToHaveSqlSchema';
import { defineDatabaseGeneratedSqlSchemaPropertiesForDomainObject } from './defineDatabaseGeneratedPropertiesForDomainObject';
import { defineSqlSchemaPropertyForDomainObjectProperty } from './defineSqlSchemaPropertyForDomainObjectProperty';
import { isNotADatabaseGeneratedProperty } from './isNotADatabaseGeneratedProperty';

export const defineSqlSchemaRelationshipForDomainObject = ({
  domainObject,
  allDomainObjects,
}: {
  domainObject: DomainObjectMetadata;
  allDomainObjects: DomainObjectMetadata[];
}): SqlSchemaToDomainObjectRelationship => {
  // figure out some relevant info
  const sqlSchemaName = snakeCase(domainObject.name); // names are in snake case

  // define the properties
  const databaseGeneratedSqlSchemaProperties =
    defineDatabaseGeneratedSqlSchemaPropertiesForDomainObject({
      domainObject,
    });
  const propertiesRelationship = [
    /// autogenerated values
    ...databaseGeneratedSqlSchemaProperties.map((sqlSchemaProperty) => {
      // see if this property was defined on the domain object (i.e., they want the value for it)
      const domainObjectProperty =
        Object.values(domainObject.properties).find(
          (property) => snakeCase(property.name) === sqlSchemaProperty.name,
        ) ?? null;

      // if so, validate that it has the correct type
      if (domainObjectProperty) {
        const expectedType = (() => {
          if (sqlSchemaProperty.name === 'id')
            return DomainObjectPropertyType.NUMBER;
          if (sqlSchemaProperty.name === 'uuid')
            return DomainObjectPropertyType.STRING;
          if (sqlSchemaProperty.name === 'created_at')
            return DomainObjectPropertyType.DATE;
          if (sqlSchemaProperty.name === 'effective_at')
            return DomainObjectPropertyType.DATE;
          if (sqlSchemaProperty.name === 'updated_at')
            return DomainObjectPropertyType.DATE;
          throw new UnexpectedCodePathDetectedError({
            reason:
              'unexpected database generated property name. this is a bug within sql-schema-generator.', // fail fast if expectations not met
            domainObjectName: domainObject.name,
            domainObjectPropertyName: sqlSchemaProperty.name,
          });
        })();
        if (domainObjectProperty.type !== expectedType)
          throw new UserInputError({
            reason: `reserved database-generated property does not match expected type. expected ${expectedType} but received ${domainObjectProperty.type}`,
            domainObjectName: domainObject.name,
            domainObjectPropertyName: domainObjectProperty.name,
          });
      }

      // return the relationship
      return {
        sqlSchema: sqlSchemaProperty,
        domainObject: domainObjectProperty,
      };
    }),

    // non autogenerated values
    ...Object.values(domainObject.properties)
      .filter(
        (domainObjectProperty) =>
          !databaseGeneratedSqlSchemaProperties
            .map(({ name }) => name)
            .includes(snakeCase(domainObjectProperty.name)), // filter out the db generated values, because we already got them above
      )
      .map((domainObjectProperty) => {
        return {
          sqlSchema: defineSqlSchemaPropertyForDomainObjectProperty({
            property: domainObjectProperty,
            domainObject,
            allDomainObjects,
          }),
          domainObject: domainObjectProperty,
        };
      }),
  ];

  // define the unique keys
  const sqlSchemaUniqueKeys = (() => {
    // define what the unique prop names are, based on what kind of object it is
    const domainObjectPropertyNamesInUniqueKey =
      domainObject.extends === DomainObjectVariant.DOMAIN_LITERAL
        ? propertiesRelationship // if its a literal, the schema is unique on all properties of the object (excluding autogenerated ones like id and uuid)
            .filter(isNotADatabaseGeneratedProperty)
            .map(
              ({ domainObject: domainObjectProperty }) =>
                domainObjectProperty.name,
            )
        : domainObject.decorations.unique; // otherwise, its unique on what was specified

    // convert the domain object prop names into sql schema prop names
    return (
      domainObjectPropertyNamesInUniqueKey?.map((domainObjectPropertyName) => {
        const sqlSchemaPropertyName = propertiesRelationship.find(
          (propertyRelationship) =>
            propertyRelationship.domainObject?.name ===
            domainObjectPropertyName,
        )?.sqlSchema.name;
        if (!sqlSchemaPropertyName)
          throw new UserInputError({
            reason: 'Unique keys must be properties of the domain object.',
            domainObjectName: domainObject.name,
            potentialSolution: ` Was there a typo defining '${
              domainObject.name
            }.unique' with key name '${domainObjectPropertyName}'? (Is it one of ${JSON.stringify(
              Object.keys(domainObject.properties),
            )}?)`,
          });
        return sqlSchemaPropertyName;
      }) ?? null
    );
  })();

  // make sure that this domain object is safe to have a schema defined for it
  assertDomainObjectIsSafeToHaveSqlSchema({ domainObject });

  // return the relationship
  return new SqlSchemaToDomainObjectRelationship({
    name: {
      sqlSchema: sqlSchemaName,
      domainObject: domainObject.name,
    },
    properties: propertiesRelationship,
    decorations: {
      alias: {
        domainObject: domainObject.decorations.alias,
      },
      unique: {
        sqlSchema: sqlSchemaUniqueKeys,
        domainObject: domainObject.decorations.unique,
      },
    },
  });
};
