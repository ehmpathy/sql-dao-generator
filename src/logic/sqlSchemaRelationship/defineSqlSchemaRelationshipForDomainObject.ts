import { snakeCase } from 'change-case';
import { DomainObjectMetadata, DomainObjectVariant } from 'domain-objects-metadata';
import { SqlSchemaToDomainObjectRelationship } from '../../domain/objects/SqlSchemaToDomainObjectRelationship';
import { assertDomainObjectIsSafeToHaveSqlSchema } from './assertDomainObjectIsSafeToHaveSqlSchema';
import { defineSqlSchemaPropertyForDomainObjectProperty } from './defineSqlSchemaPropertyForDomainObjectProperty';

export const defineSqlSchemaRelationshipForDomainObject = ({
  domainObject,
  allDomainObjects,
}: {
  domainObject: DomainObjectMetadata;
  allDomainObjects: DomainObjectMetadata[];
}) => {
  // define the name
  const sqlSchemaName = snakeCase(domainObject.name); // names are in snake case

  // define the properties
  const propertiesRelationship = Object.values(domainObject.properties)
    .filter((property) => !['id', 'uuid'].includes(property.name))
    .map((property) => {
      return {
        sqlSchema: defineSqlSchemaPropertyForDomainObjectProperty({ property, domainObject, allDomainObjects }),
        domainObject: property,
      };
    });

  // define the unique keys
  const sqlSchemaUniqueKeys = (() => {
    // define what the unique prop names are, based on what kind of object it is
    const domainObjectPropertyNamesInUniqueKey =
      domainObject.extends === DomainObjectVariant.DOMAIN_VALUE_OBJECT
        ? Object.keys(domainObject.properties).filter((key) => !['id', 'uuid'].includes(key)) // if its a value object, the schema is unique on all properties of the object (excluding autogenerated ones like id and uuid)
        : domainObject.decorations.unique;

    // convert the domain object prop names into sql schema prope names
    return (
      domainObjectPropertyNamesInUniqueKey?.map((domainObjectPropertyName) => {
        if (domainObjectPropertyName === 'uuid') return 'uuid'; // handle edge case where object not unique on any natural keys
        const sqlSchemaPropertyName = propertiesRelationship.find(
          (propertyRelationship) => propertyRelationship.domainObject.name === domainObjectPropertyName,
        )?.sqlSchema.name;
        if (!sqlSchemaPropertyName)
          throw new Error(
            `Unique keys must be properties of the domain object. Was there a typo defining '${
              domainObject.name
            }.unique' with key name '${domainObjectPropertyName}'. (Is it one of ${JSON.stringify(
              Object.keys(domainObject.properties),
            )}?)`,
          );
        return sqlSchemaPropertyName;
      }) ?? null
    );
  })();

  // make sure that this domain object is safe to have a schema defined for it
  assertDomainObjectIsSafeToHaveSqlSchema({ domainObject });

  // return the relationship
  return new SqlSchemaToDomainObjectRelationship({
    name: {
      sqlSchema: sqlSchemaName,
      domainObject: domainObject.name,
    },
    properties: propertiesRelationship,
    decorations: {
      unique: {
        sqlSchema: sqlSchemaUniqueKeys,
        domainObject: domainObject.decorations.unique,
      },
    },
  });
};
