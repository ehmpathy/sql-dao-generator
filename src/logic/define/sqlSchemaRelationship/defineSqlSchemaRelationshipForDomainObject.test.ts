import { DomainObjectMetadata, DomainObjectPropertyType, DomainObjectVariant } from 'domain-objects-metadata';
import { createExampleDomainObjectMetadata } from '../../__test_assets__/createExampleDomainObject';

import { defineSqlSchemaRelationshipForDomainObject } from './defineSqlSchemaRelationshipForDomainObject';

describe('defineSqlSchemarelationshipForDomainObject', () => {
  it('should look right for a domain-value-object', () => {
    const relationship = defineSqlSchemaRelationshipForDomainObject({
      domainObject: new DomainObjectMetadata({
        name: 'PreciseGeocode',
        extends: DomainObjectVariant.DOMAIN_VALUE_OBJECT,
        properties: {
          id: { name: 'id', type: DomainObjectPropertyType.NUMBER },
          uuid: { name: 'uuid', type: DomainObjectPropertyType.STRING },
          latitude: { name: 'latitude', type: DomainObjectPropertyType.NUMBER },
          longitude: { name: 'longitude', type: DomainObjectPropertyType.NUMBER },
        },
        decorations: {
          unique: null,
          updatable: null,
        },
      }),
      allDomainObjects: [],
    });
    expect(relationship.name.sqlSchema).toEqual('precise_geocode'); // should be snake case
    expect(relationship.properties.length).toEqual(2); // sanity check
    expect(relationship.decorations.unique.domainObject).toEqual(null); // it wasn't defined, since domain value object
    expect(relationship.decorations.unique.sqlSchema).toEqual(['latitude', 'longitude']); // should be all of the properties, since domain value object
    expect(relationship).toMatchSnapshot();
  });
  it('should look right for a domain-entity', () => {
    const relationship = defineSqlSchemaRelationshipForDomainObject({
      domainObject: new DomainObjectMetadata({
        name: 'TrainCarriage',
        extends: DomainObjectVariant.DOMAIN_ENTITY,
        properties: {
          id: { name: 'id', type: DomainObjectPropertyType.NUMBER, required: false },
          uuid: { name: 'uuid', type: DomainObjectPropertyType.STRING, required: false },
          cin: { name: 'cin', type: DomainObjectPropertyType.STRING, required: true },
          carries: {
            name: 'carries',
            type: DomainObjectPropertyType.ENUM,
            of: ['PASSENGER', 'FREIGHT'],
            required: true,
          },
          capacity: { name: 'capacity', type: DomainObjectPropertyType.NUMBER, nullable: true },
          manufacturer: {
            name: 'manufacturer',
            type: DomainObjectPropertyType.REFERENCE,
            of: { name: 'TrainManufacturer', extends: DomainObjectVariant.DOMAIN_VALUE_OBJECT },
          },
        },
        decorations: {
          unique: ['cin'],
          updatable: [],
        },
      }),
      allDomainObjects: [],
    });
    expect(relationship.name.sqlSchema).toEqual('train_carriage'); // should be snake case
    expect(relationship.properties.length).toEqual(4); // only includes the non-auto-generated ones
    expect(relationship.decorations.unique.sqlSchema).toEqual(['cin']); // sanity check
    expect(relationship).toMatchSnapshot();
  });
  it('should look right for a domain-event', () => {
    const relationship = defineSqlSchemaRelationshipForDomainObject({
      domainObject: new DomainObjectMetadata({
        name: 'TrainLocatedEvent',
        extends: DomainObjectVariant.DOMAIN_EVENT,
        properties: {
          id: { name: 'id', type: DomainObjectPropertyType.NUMBER, required: false },
          trainUuid: { name: 'trainUuid', type: DomainObjectPropertyType.STRING, required: true },
          occurredAt: { name: 'occurredAt', type: DomainObjectPropertyType.DATE, required: true },
          geocode: {
            name: 'geocode',
            type: DomainObjectPropertyType.REFERENCE,
            of: {
              name: 'Geocode',
              extends: DomainObjectVariant.DOMAIN_VALUE_OBJECT,
            },
            required: true,
          },
        },
        decorations: {
          unique: ['trainUuid', 'occurredAt'],
          updatable: [],
        },
      }),
      allDomainObjects: [
        { ...createExampleDomainObjectMetadata(), name: 'Train', extends: DomainObjectVariant.DOMAIN_ENTITY },
      ],
    });
    expect(relationship.name.sqlSchema).toEqual('train_located_event'); // should be snake case
    expect(relationship.properties.length).toEqual(3); // only includes the non-auto-generated ones
    expect(relationship.decorations.unique.sqlSchema).toEqual(['train_id', 'occurred_at']); // notice that `trainUuid` was converted to `train_id`! (since the sql column is called `train_id`, since it references a `train`)
    expect(relationship).toMatchSnapshot();
  });
});
