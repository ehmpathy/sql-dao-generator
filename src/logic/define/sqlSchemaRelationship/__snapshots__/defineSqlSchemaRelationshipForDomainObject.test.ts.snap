// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defineSqlSchemarelationshipForDomainObject should look right for a domain-entity 1`] = `
SqlSchemaToDomainObjectRelationship {
  "decorations": {
    "unique": {
      "domainObject": [
        "cin",
      ],
      "sqlSchema": [
        "cin",
      ],
    },
  },
  "name": {
    "domainObject": "TrainCarriage",
    "sqlSchema": "train_carriage",
  },
  "properties": [
    {
      "domainObject": {
        "name": "id",
        "required": false,
        "type": "NUMBER",
      },
      "sqlSchema": SqlSchemaPropertyMetadata {
        "isArray": false,
        "isDatabaseGenerated": true,
        "isNullable": false,
        "isUpdatable": false,
        "name": "id",
        "reference": null,
      },
    },
    {
      "domainObject": {
        "name": "uuid",
        "required": false,
        "type": "STRING",
      },
      "sqlSchema": SqlSchemaPropertyMetadata {
        "isArray": false,
        "isDatabaseGenerated": true,
        "isNullable": false,
        "isUpdatable": false,
        "name": "uuid",
        "reference": null,
      },
    },
    {
      "domainObject": null,
      "sqlSchema": SqlSchemaPropertyMetadata {
        "isArray": false,
        "isDatabaseGenerated": true,
        "isNullable": false,
        "isUpdatable": false,
        "name": "created_at",
        "reference": null,
      },
    },
    {
      "domainObject": null,
      "sqlSchema": SqlSchemaPropertyMetadata {
        "isArray": false,
        "isDatabaseGenerated": true,
        "isNullable": false,
        "isUpdatable": false,
        "name": "effective_at",
        "reference": null,
      },
    },
    {
      "domainObject": null,
      "sqlSchema": SqlSchemaPropertyMetadata {
        "isArray": false,
        "isDatabaseGenerated": true,
        "isNullable": false,
        "isUpdatable": false,
        "name": "updated_at",
        "reference": null,
      },
    },
    {
      "domainObject": {
        "name": "cin",
        "required": true,
        "type": "STRING",
      },
      "sqlSchema": SqlSchemaPropertyMetadata {
        "isArray": false,
        "isDatabaseGenerated": false,
        "isNullable": false,
        "isUpdatable": false,
        "name": "cin",
        "reference": null,
      },
    },
    {
      "domainObject": {
        "name": "carries",
        "of": [
          "PASSENGER",
          "FREIGHT",
        ],
        "required": true,
        "type": "ENUM",
      },
      "sqlSchema": SqlSchemaPropertyMetadata {
        "isArray": false,
        "isDatabaseGenerated": false,
        "isNullable": false,
        "isUpdatable": false,
        "name": "carries",
        "reference": null,
      },
    },
    {
      "domainObject": {
        "name": "capacity",
        "nullable": true,
        "type": "NUMBER",
      },
      "sqlSchema": SqlSchemaPropertyMetadata {
        "isArray": false,
        "isDatabaseGenerated": false,
        "isNullable": true,
        "isUpdatable": true,
        "name": "capacity",
        "reference": null,
      },
    },
    {
      "domainObject": {
        "name": "manufacturer",
        "of": {
          "extends": "DomainValueObject",
          "name": "TrainManufacturer",
        },
        "type": "REFERENCE",
      },
      "sqlSchema": SqlSchemaPropertyMetadata {
        "isArray": false,
        "isDatabaseGenerated": false,
        "isNullable": false,
        "isUpdatable": false,
        "name": "manufacturer_id",
        "reference": SqlSchemaReferenceMetadata {
          "method": "DIRECT_BY_NESTING",
          "of": DomainObjectReferenceMetadata {
            "extends": "DomainValueObject",
            "name": "TrainManufacturer",
          },
        },
      },
    },
  ],
}
`;

exports[`defineSqlSchemarelationshipForDomainObject should look right for a domain-event 1`] = `
SqlSchemaToDomainObjectRelationship {
  "decorations": {
    "unique": {
      "domainObject": [
        "trainUuid",
        "occurredAt",
      ],
      "sqlSchema": [
        "train_id",
        "occurred_at",
      ],
    },
  },
  "name": {
    "domainObject": "TrainLocatedEvent",
    "sqlSchema": "train_located_event",
  },
  "properties": [
    {
      "domainObject": {
        "name": "id",
        "required": false,
        "type": "NUMBER",
      },
      "sqlSchema": SqlSchemaPropertyMetadata {
        "isArray": false,
        "isDatabaseGenerated": true,
        "isNullable": false,
        "isUpdatable": false,
        "name": "id",
        "reference": null,
      },
    },
    {
      "domainObject": null,
      "sqlSchema": SqlSchemaPropertyMetadata {
        "isArray": false,
        "isDatabaseGenerated": true,
        "isNullable": false,
        "isUpdatable": false,
        "name": "uuid",
        "reference": null,
      },
    },
    {
      "domainObject": null,
      "sqlSchema": SqlSchemaPropertyMetadata {
        "isArray": false,
        "isDatabaseGenerated": true,
        "isNullable": false,
        "isUpdatable": false,
        "name": "created_at",
        "reference": null,
      },
    },
    {
      "domainObject": {
        "name": "trainUuid",
        "required": true,
        "type": "STRING",
      },
      "sqlSchema": SqlSchemaPropertyMetadata {
        "isArray": false,
        "isDatabaseGenerated": false,
        "isNullable": false,
        "isUpdatable": false,
        "name": "train_id",
        "reference": SqlSchemaReferenceMetadata {
          "method": "IMPLICIT_BY_UUID",
          "of": DomainObjectReferenceMetadata {
            "extends": "DomainEntity",
            "name": "Train",
          },
        },
      },
    },
    {
      "domainObject": {
        "name": "occurredAt",
        "required": true,
        "type": "DATE",
      },
      "sqlSchema": SqlSchemaPropertyMetadata {
        "isArray": false,
        "isDatabaseGenerated": false,
        "isNullable": false,
        "isUpdatable": false,
        "name": "occurred_at",
        "reference": null,
      },
    },
    {
      "domainObject": {
        "name": "geocode",
        "of": {
          "extends": "DomainValueObject",
          "name": "Geocode",
        },
        "required": true,
        "type": "REFERENCE",
      },
      "sqlSchema": SqlSchemaPropertyMetadata {
        "isArray": false,
        "isDatabaseGenerated": false,
        "isNullable": false,
        "isUpdatable": false,
        "name": "geocode_id",
        "reference": SqlSchemaReferenceMetadata {
          "method": "DIRECT_BY_NESTING",
          "of": DomainObjectReferenceMetadata {
            "extends": "DomainValueObject",
            "name": "Geocode",
          },
        },
      },
    },
  ],
}
`;

exports[`defineSqlSchemarelationshipForDomainObject should look right for a domain-value-object 1`] = `
SqlSchemaToDomainObjectRelationship {
  "decorations": {
    "unique": {
      "domainObject": null,
      "sqlSchema": [
        "latitude",
        "longitude",
      ],
    },
  },
  "name": {
    "domainObject": "PreciseGeocode",
    "sqlSchema": "precise_geocode",
  },
  "properties": [
    {
      "domainObject": {
        "name": "id",
        "type": "NUMBER",
      },
      "sqlSchema": SqlSchemaPropertyMetadata {
        "isArray": false,
        "isDatabaseGenerated": true,
        "isNullable": false,
        "isUpdatable": false,
        "name": "id",
        "reference": null,
      },
    },
    {
      "domainObject": {
        "name": "uuid",
        "type": "STRING",
      },
      "sqlSchema": SqlSchemaPropertyMetadata {
        "isArray": false,
        "isDatabaseGenerated": true,
        "isNullable": false,
        "isUpdatable": false,
        "name": "uuid",
        "reference": null,
      },
    },
    {
      "domainObject": null,
      "sqlSchema": SqlSchemaPropertyMetadata {
        "isArray": false,
        "isDatabaseGenerated": true,
        "isNullable": false,
        "isUpdatable": false,
        "name": "created_at",
        "reference": null,
      },
    },
    {
      "domainObject": {
        "name": "latitude",
        "type": "NUMBER",
      },
      "sqlSchema": SqlSchemaPropertyMetadata {
        "isArray": false,
        "isDatabaseGenerated": false,
        "isNullable": false,
        "isUpdatable": false,
        "name": "latitude",
        "reference": null,
      },
    },
    {
      "domainObject": {
        "name": "longitude",
        "type": "NUMBER",
      },
      "sqlSchema": SqlSchemaPropertyMetadata {
        "isArray": false,
        "isDatabaseGenerated": false,
        "isNullable": false,
        "isUpdatable": false,
        "name": "longitude",
        "reference": null,
      },
    },
  ],
}
`;

exports[`defineSqlSchemarelationshipForDomainObject should look right for another domain-value-object, one with multi word property names 1`] = `
SqlSchemaToDomainObjectRelationship {
  "decorations": {
    "unique": {
      "domainObject": null,
      "sqlSchema": [
        "role",
        "external_id",
      ],
    },
  },
  "name": {
    "domainObject": "ChatParticipant",
    "sqlSchema": "chat_participant",
  },
  "properties": [
    {
      "domainObject": {
        "name": "id",
        "type": "NUMBER",
      },
      "sqlSchema": SqlSchemaPropertyMetadata {
        "isArray": false,
        "isDatabaseGenerated": true,
        "isNullable": false,
        "isUpdatable": false,
        "name": "id",
        "reference": null,
      },
    },
    {
      "domainObject": {
        "name": "uuid",
        "type": "STRING",
      },
      "sqlSchema": SqlSchemaPropertyMetadata {
        "isArray": false,
        "isDatabaseGenerated": true,
        "isNullable": false,
        "isUpdatable": false,
        "name": "uuid",
        "reference": null,
      },
    },
    {
      "domainObject": null,
      "sqlSchema": SqlSchemaPropertyMetadata {
        "isArray": false,
        "isDatabaseGenerated": true,
        "isNullable": false,
        "isUpdatable": false,
        "name": "created_at",
        "reference": null,
      },
    },
    {
      "domainObject": {
        "name": "role",
        "type": "STRING",
      },
      "sqlSchema": SqlSchemaPropertyMetadata {
        "isArray": false,
        "isDatabaseGenerated": false,
        "isNullable": false,
        "isUpdatable": false,
        "name": "role",
        "reference": null,
      },
    },
    {
      "domainObject": {
        "name": "externalId",
        "type": "STRING",
      },
      "sqlSchema": SqlSchemaPropertyMetadata {
        "isArray": false,
        "isDatabaseGenerated": false,
        "isNullable": false,
        "isUpdatable": false,
        "name": "external_id",
        "reference": null,
      },
    },
  ],
}
`;
